src/Microsoft.Data.Sqlite.Core/SqliteConnection.cs

```diff
@@ -297,6 +299,352 @@ public virtual void CreateCollation<T>(string name, T state, Func<T, string, str
+
+        private void CreateAggregateFunction(
+            string name,
+            int numberOfArguments,
+            delegate_function_aggregate_step stepFunction,
+            delegate_function_aggregate_final finalFunction,
+            [CallerMemberName] string callerName = "")
+        {
+            if (string.IsNullOrEmpty(name))
+            {
+                throw new ArgumentNullException(nameof(name));
+            }
+            if (State != ConnectionState.Open)
+            {
+                throw new InvalidOperationException(Resources.CallRequiresOpenConnection(callerName));
+            }
+
+            var rc = raw.sqlite3_create_function(_db, name, numberOfArguments, null, stepFunction, finalFunction);
+            SqliteException.ThrowExceptionForRC(rc, _db);
+        }
+
+        private T GetArgument<T>(sqlite3_value value)
+        {
+            var type = typeof(T).UnwrapNullableType().UnwrapEnumType();
+            if (type == typeof(DBNull))
+            {
+                return (T)(object)DBNull.Value;
+            }
+            if (type == typeof(object))
+            {
+                return (T)GetArgumentObject(value);
+            }
+
+            return SqliteDataReader.GetValue<T>(
+                type,
+                () => raw.sqlite3_value_type(value),
+                () => raw.sqlite3_value_int64(value),
+                () => raw.sqlite3_value_double(value),
+                () => raw.sqlite3_value_text(value),
+                () => raw.sqlite3_value_blob(value),
+                () => throw new InvalidCastException());
+        }
+
+        private static object GetArgumentObject(sqlite3_value value)
+        {
+            var sqliteType = raw.sqlite3_value_type(value);
+            switch (sqliteType)
+            {
+                case raw.SQLITE_INTEGER:
+                    return raw.sqlite3_value_int64(value);
+                case raw.SQLITE_FLOAT:
+                    return raw.sqlite3_value_double(value);
+                case raw.SQLITE_BLOB:
+                    return raw.sqlite3_value_blob(value);
+                case raw.SQLITE_TEXT:
+                    return raw.sqlite3_value_text(value);
+                default:
+                    throw new InvalidCastException();
+            }
+        }
+
+        private static object GetArgumentsArray(sqlite3_value[] args)
+        {
+            if (args == null)
+            {
+                throw new ArgumentNullException(nameof(args)); // unknown if this could ever happen
+            }
+
+            var array = new object[args.Length];
+            for (var i = 0; i < args.Length; i++)
+            {
+                array[i] = GetArgumentObject(args[i]);
+            }
+
+            return array;
+        }
+
+        private void SetFunctionResult(object result, sqlite3_context ctx)
+        {
+            var type = result.GetType().UnwrapNullableType().UnwrapEnumType();
+
+            SqliteParameter.CoerceValue(
```

Comment:
With the newest version this is refactored into new classes which do the value storage. Since SqliteParameter uses this also for rebinding the value to a new statement/index the base class is a bit more complex than needed for the result storage.